File: python/42_climbing_stairs.py
"""
Question:
You are climbing a staircase. It takes n steps to reach to the top.
Each time you can either climb 1 or 2 steps. In how many distinct ways can
you climb to the top?
"""


class Solution:
    def climbStairs(self, n: 'int') -> 'int':
        if n == 1:
            return 1
        dp = [0] * (n + 1)
        dp[1] = 1
        dp[2] = 2
        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[n]

File: cpp/42_climbing_stairs.cpp
// // https://leetcode.com/problems/climbing-stairs/
// You are climbing a staircase.It takes n steps to reach the top.

//     Each time you can either climb 1 or
//     2 steps.In how many distinct ways can you climb to the top
//     ?

class Solution {
   public:
    int climbStairs(int n) {
        if (n <= 2) return n;
        int a = 2, b = 1, res;
        for (int i = 3; i <= n; i++) {
            res = a + b;
            b = a;
            a = res;
        }
        return res;
    }
};